cpp/src/endpoint.cpp

```diff
@@ -67,12 +66,17 @@ Endpoint::Endpoint(std::shared_ptr<Component> workerOrListener,
 
   setParent(workerOrListener);
 
-  _callbackData = std::make_unique<ErrorCallbackData>(this, _inflightRequests, worker);
+  _callbackData = std::make_unique<ErrorCallbackData>(_inflightRequests, worker);
 
-  params->err_mode =
-    (endpointErrorHandling ? UCP_ERR_HANDLING_MODE_PEER : UCP_ERR_HANDLING_MODE_NONE);
-  params->err_handler.cb  = Endpoint::errorCallback;
-  params->err_handler.arg = _callbackData.get();
+  if (endpointErrorHandling) {
+    params->err_mode        = UCP_ERR_HANDLING_MODE_PEER;
+    params->err_handler.cb  = Endpoint::errorCallback;
+    params->err_handler.arg = _callbackData.get();
+  } else {
+    params->err_mode        = UCP_ERR_HANDLING_MODE_NONE;
+    params->err_handler.cb  = nullptr;
```

Comment:
I agree the name has become misleading, I initially added it exclusively as means to communicate with the error callback, but now it is used to store any information that _may_ require to be shared with the error callback, such as the `_callbackData.closing` flag which is used in various parts of the code inside and out of the `Endpoint::errorCallback()`. I think we could handle it in two ways:

1. Rename `ErrorCallbackData` to something that is more appropriate (do you have any naming suggestions?); or
1. Duplicate things like `_callbackData.closing`, but I personally hate this because it's certain to become a problem when someone makes a change that accidentally missing synchronizing.

Reply:
tl;dr: maybe a name change is OK if we can think of a good one, otherwise, do nothing,

Of the two, I definitely (like you) don't like (2). We could just call it `EndpointData`? But then the question arises, in what circumstances do I need this data where I don't have the endpoint? IOW, if the `Endpoint` directly provided the information that we need to pass in to `ucp_ep_create` (say), could we just use it directly? Or does that bring all sorts of problems with destruction order etc...

Reply:
> But then the question arises, in what circumstances do I need this data where I don't have the endpoint?

This really shouldn't be necessary as the `ErrorCallbackData` is itself owned by the `Endpoint`. I haven't been able to determine why it may happen at times that the endpoint looks like it's been destroyed, but I'm thinking it may have something to do with generic callbacks.

I think one of the issues where we're seeing segfaults elsewhere are the generic callbacks, specifically after a timeout occurred, the [`Endpoint::closeBlocking`](https://github.com/rapidsai/ucxx/blob/b117fda698c76a3685b5b7d14008609eb8d1d8a9/cpp/src/endpoint.cpp#L226-L243) is one such example, we register the callback with the worker and wait for it, if it doesn't complete we simply get to the next iteration of the loop. The problem is those objects are all locally-scoped and we never deregister from the worker, so if/when the worker actually attempts to execute the callback the function and objects do not exist anymore, leading to potential segfaults, which is definitely something I'll try to address in a follow-up PR within the next few days.

> IOW, if the `Endpoint` directly provided the information that we need to pass in to `ucp_ep_create` (say), could we just use it directly? Or does that bring all sorts of problems with destruction order etc...

This is an excellent point you're making, we don't really need the `ErrorCallbackData` provided we can pass a pointer to `this` directly to `errorCallback()`, which we now can do thanks to the `create()` method. With that said I've now attempted to remove `ErrorCallbackData` and pass `this` directly to a `errorCallback` function which is now a `friend` of `ucxx::Endpoint` and can thus operate directly on the `Endpoint` instance itself. However, I'm not confident this won't cause the original issue I'm trying to address with this PR so before we merge it I'll have a stress test running locally for a day or so and will trigger CI multiple times to see if we can reproduce the segfault.

Reply:
The change is in https://github.com/rapidsai/ucxx/pull/227/commits/ade6aff99c906f3fafdcf11120ab66d82e6b7360 .

Reply:
As discussed offline, unfortunately https://github.com/rapidsai/ucxx/commit/ade6aff99c906f3fafdcf11120ab66d82e6b7360 has reintroduced the segfault which is likely rooted at the generic callbacks, therefore I reverted it and will reintroduced that change along with generic callback fixes in a follow-up PR.